src/main/haskell/kore/src/Kore/Step/Strategy.hs

```diff
@@ -0,0 +1,323 @@
+{-|
+Module      : Kore.Step.Strategy
+Description : Strategies for pattern rewriting
+Copyright   : (c) Runtime Verification, 2018
+License     : NCSA
+Maintainer  : thomas.tuegel@runtimeverification.com
+-}
+module Kore.Step.Strategy
+    ( -- * Strategies
+      Strategy
+    , and
+    , all
+    , or
+    , try
+    , any
+    , many
+    , some
+    , apply
+    , step
+    , stuck
+      -- * Running strategies
+    , runStrategy
+    , pickLongest
+    , pickStuck
+      -- * Re-exports
+    , module Data.Limit
+    , Tree (..)
+    ) where
+
+import           Data.Bifunctor
+                 ( first )
+import           Data.List.NonEmpty
+                 ( NonEmpty (..) )
+import           Data.Semigroup
+import           Data.Tree
+                 ( Tree )
+import qualified Data.Tree as Tree
+import           Prelude hiding
+                 ( all, and, any, or )
+
+import Control.Monad.Counter
+import Data.Limit
+
+{- | An execution strategy.
+
+  @Strategy prim@ represents a strategy for execution by applying rewrite axioms
+  of type @prim@.
+
+ -}
+data Strategy prim where
+
+    -- The recursive arguments of these constructors are /intentionally/ lazy to
+    -- allow strategies to loop.
+
+    And :: Strategy prim -> Strategy prim -> Strategy prim
```

Comment:
Could you consider describing these cases? Since you didn't export them, the style guide does not require it, but it may be useful, at least for some of them.

And and Or may be obvious-ish, but Step is less so, and Apply is fairly confusing at first sight (e.g. why does it take a 'prim' and a 'Strategy' to produce a 'Strategy'? Why isn't called 'Sequence' or 'ApplySequence' or something similar? Why aren't 'Apply' and 'Sequence' independent concepts?).

Reply:
@virgil-serbanuta I added more documentation about `Strategy` which I believe addresses your questions here. However, I was not entirely sure if your questions were intended to direct me in writing documentation, or if you actually wanted to discuss the design here. Either way, I think the documentation addresses your questions, but I am happy to discuss more.